### ISSUE TITLE
[expfactory] template modifiers

### ISSUE THREAD

**Issue by vsoch** [FelixHenninger/lab.js#18]
Hey @FelixHenninger ! I've been testing the labjs integration, and there are a few tweaks to make so that the experiment will function properly! 

 **jquery**

should be added in the basic header of index.html, like this -->
https://github.com/expfactory/builder-labjs/blob/master/experiments/stroop-task/index.html#L7

I found the expfactory modifier script, https://github.com/FelixHenninger/lab.js/blob/master/packages/builder/src/logic/io/export/modifiers/expfactory.js and maybe we could have it added with a function in here? I'm not familiar with the setup so wanted to discuss with you. 

**response**
This isn't required, but it's a suggestion to show the response even if it's not perfect, because it's needed for debugging. I added an "else" that prints the response if it's not ok --> https://github.com/expfactory/builder-labjs/blob/master/experiments/stroop-task/index.html#L46

**bug in transmit**
When I submit the data to /save, I get a 400 BAD REQUEST. I think this is because of the format of the data that is being submit? It is coming from `lab.plugins.Transmit`:

```javascript
    {
      "type": "lab.plugins.Transmit",
      "url": "/save",
      "updates": {
        "staging": false,
        "headers": {"X-CSRFToken": csrf_token}
      },
      "callbacks": {
        "full": function full(response) {
        if (response && response.ok) {
          console.log(response);
          window.location = '/next';
        } else {
          console.log(response);          
        }
      }
      }
```
And this I think is generated from here? https://github.com/FelixHenninger/lab.js/blob/master/packages/library/src/plugins/transmit.js#L3

The above doesn't show me how the submit is being done, and I'd like to better understand the format of the data being transmit to properly debug. For expfactory experiments, I typically just submit json to the endpoint, and I use ajax since the header for the csrf_token is added that way, like this:

```javascript
        $.ajax({
                         type: "POST",
                         url: '/save',
                         data: { "data": data },
                         success: function(){ document.location = "/next" },
                         dataType: "application/json",
                         // Endpoint not running, local save
                         error: function(err) {
                             if (err.status == "200"){
                                 document.location = "/next";
                             } else {
                                 // if error, assume local save
                                 jsPsych.data.localSave('breath-counting-experiment_results.csv', 'csv');
                             }
                         }
                     });
```

We need some tweaking somewhere to get this finished up!

And exciting things to share (that will help you to possibly debug this issue!) I have the basic "expfactory labjs builder" finished up, and once the experiment is fully working I can do the little writeup. It works that you literally export the LabJS experiment, and drop it in a folder, and you're done :) Here is the repo:

 - Here is the repo --> https://github.com/expfactory/builder-labjs
 - Here is the build and deploy on Circle --> https://circleci.com/workflow-run/7913bb01-4b2a-44be-94a6-0439dbc6bfae
 - And deployment (Docker Hub) is here -->  https://hub.docker.com/r/vanessa/expfactory-stroop/

So - to run this experiment on your computer with docker, just do:

```bash
docker run -d -p 80:80 --name stroop vanessa/expfactory-stroop start
```
And then open to [http://127.0.0.1](http://127.0.0.1). The experiment will proceed and what I'm debugging now (the things above to discuss) are the submission of the data, what I see in the console is:

![image](https://user-images.githubusercontent.com/814322/42724473-d8c2daaa-8727-11e8-884f-1d266979efd2.png)

and I'm trying to debug and need your help! It could either be that the data format is off, or the passing of the cookie isn't working. If we can talk about the above I can get a better sense of this - I'm super excited to get this underway!

**Comment by vsoch** [comment]
A quick update with more debugging - I removed the ajaxSetup and it didn't alter the error, so that's not it!

**Comment by vsoch** [comment]
I'm going out for a run so I'll leave it here for you to take a look - I'd like to propose that we figure out how to fix this up, and then I'll finish the writeup, and add the docs to the LabJS base (I started this but obviously need to get it fully working first, haha). Then once the post is out we can get feedback from the community about ideas for the experiment library / "where the metadata goes." The Circle CI orbs that I mentioned are a (future) option to make this even more templatized, although theoretically it's not needed.

**Comment by vsoch** [comment]
ok sounds good!

**Comment by FelixHenninger** [comment]
Hej again,

I've been trying out things, and the expfactory image is failing for me, do you have an idea what I'm doing wrong? I'm running the docker commands in the folder `expfactory-test`, with a nested folder `data`, which in turn contains the stroop task in `data/stroop-task`.

```
docker run -v $PWD/data:/data vanessa/expfactory-builder build stroop /data/stroop-task
```

works fine for me, but 

```
docker build --no-cache -t expfactory/experiments .
```

in the `data` directory isn't at all happy ([build.log](https://github.com/FelixHenninger/lab.js/files/2195067/build.log)). I'm assuming that I have an old version of an image somewhere, but I've pulled the latest `expfactory-builder`.

Sorry for the trouble -- any hints would be super-appreciated! 

-Felix

**Comment by vsoch** [comment]
Oh! so the testing happens on the continuous integration server, so you shouldn't need to do it locally. Just try running the container with the command that I showed you above and it will pull and run from

**Comment by FelixHenninger** [comment]
Wow, thanks for your quick response! That makes sense, and I'll do just that -- sorry, I thought debugging might be easier if I do things locally. Please give me a sec!

**Comment by vsoch** [comment]
Just run this:

docker run -d -p 80:80 --name stroop vanessa/expfactory-stroop start

**Comment by vsoch** [comment]
Sure thing! I found for debugging it's easier to just commit and push to GitHub and then have the image build automatically and push to dockerhub and then use it. so theoretically if you have forked my repo you could also just connect to Docker Hub and Define the container name and Docker credentials in the environment settings and then it will do the same for you.

**Comment by FelixHenninger** [comment]
Ok, so this works for me (sorry for going the convoluted route immediately -- I'm not yet fully used to the fancy containerized/CI workflow!). So from within the experiment, I can get the data to be saved via 

```
$.ajax({
  type: "POST",
  url: '/save',
  data: { "data": study.options.datastore.exportJson() },
  dataType: "application/json",
  success: function() { console.log('success') },
  error: function(err) { console.log('error', err) }
});
```

which, oddly, only ever triggers the error handler, even though the response is `200` and contains `{"success": true}` -- but then again you anticipate that in your code above.

The central difference to the `lab.js` data submission seems to be that jQuery sends the information form-encoded. Is that intentional? I'm guessing that that is also the reason that the server rejects the request; I'll look into that now.

**Comment by vsoch** [comment]
okay let's do this - first I'll (manually) add this to the container, build with CircleCI, and then pull and test on my own locally. This is where I can debug (it is telling you it's an error? It might not be I think? I can look into this) When we have something working and I've verified, I can show you the index.html that was used for the experiment and then we can talk about how to "get that thing exported" from the LabJS builder. Does that sound like a plan?

Apologies for the delay - went on a longer run and trip out and I'm back at keyboard!

**Comment by vsoch** [comment]
oh hmm. How do I get this:

```javascript
$.ajax({
  type: "POST",
  url: '/save',
  data: { "data": study.options.datastore.exportJson() },
  dataType: "application/json",
  success: function() { console.log('success') },
  error: function(err) { console.log('error', err) }
});
```
Into this format?

```javascript
    {
      "type": "lab.plugins.Transmit",
      "url": "/save",
      "updates": {
        "staging": false ,
        "headers": {"X-CSRFToken": "{{ csrf_token() }}"}
      },
      "callbacks": {
        "full": function full(response) {
        if (response && response.ok) {
          console.log(response);
          window.location = '/next';
        } else {
          console.log(response);
        }
      }
      }
    }
```
E.g., how would that data structure for the study look to have the ajax post? This all comes from the experiment that is exported from LabJS, and the only thing I've added is the little snippet to add the csrf token via ajax header (which looks like it isn't needed because you already added the header!)

**Comment by FelixHenninger** [comment]
Hej, welcome back, your timing is perfect -- I just figured out what I was doing wrong all the time! I think I've gotten things working without jQuery, but that will require a change to the `lab.js` code. The issue was indeed that jQuery treats the transmitted JSON data as a form, and puts things in different fields etc.; we'll need to add that as an option in the `Transmit` plugin in `lab.js`.

**Comment by vsoch** [comment]
okay great that you found the issue! How can I help? Should I wait for a different export to test with?

**Comment by vsoch** [comment]
And I feel silly I thought that "Hej" was a typo but now it would be too many times for chance! How do you pronounce that, is it supposed to be like "Hi" ?

**Comment by FelixHenninger** [comment]
Give me a moment, I'll need to figure things out! In the long term, it might be worth thinking about creating a small `expfactory` library that takes the data and handles the submission; that way, you could decouple the container internals from the experiment logic.

Oh, and regarding 'hej', it's [the same (in pronunciation and meaning) as 'hey'](https://www.youtube.com/watch?v=x688wjoMjfI) -- sorry for the confusion, that's a quirk of mine I guess.

**Comment by vsoch** [comment]
That's a really good idea! If I gave a try at starting something, could you help / advise? I am not so prolific with Javascript, especially (is it called ES6?) I had wanted to not put any requirements on the experiments other than "submit data to this endpoint" but I'm of course realizing it's not so simple like that. Do you think if there was an expfactory.js others would be ok to use it to submit their data?

**Comment by vsoch** [comment]
oh haha, it's literally just "Hey!" I'm so dense sometimes :P

**Comment by FelixHenninger** [comment]
Oh, no worries, sorry for the confusion! And sure, I'd be thrilled to help with a small library! I don't think it would take a lot of work -- for starters, you could wrap something like the following code (or I could do it for you, if you'd like):

```
const formData = new Form()
formData.append('data', 'data from the experiment, as string')

fetch('/save', {
  method: 'post',
  headers: {
    'Accept': 'application/json',
    'Content-Type': 'application/json',
    'X-CSRFToken': csrf_token,
  },
  body: formData,
  credentials: 'include',
}).then(response => {
  if (response && response.ok) {
    window.location = '/next';
  } else {
    // Server signalled problem, e.g. error code
    console.log(response);
  }
}).catch(error => {
  // deal with transmission error, e.g. retry
  console.log(error)
})
```

You'd probably want to include a [`fetch` polyfill](https://github.com/github/fetch) for [browsers that don't support it yet](https://caniuse.com/#search=fetch) (pretty much only IE11 at this point), but something like this should do the trick -- and the studies wouldn't have to load all of jQuery if they don't need it (I don't know whether jsPsych has fully removed the dependency yet).

**Comment by vsoch** [comment]
Oh haha, I just spent some time on something that is pretty terrible compared to this, I'll post it soon!

Is there any reason it has to be form data? Is that just a convention / holder for a general data structure? I want to generally allow any json-ifyable object to be stored (outside of the constrains of formy type things).

**Comment by vsoch** [comment]
Here is what I have so far - I think maybe the (rather bad way) I implemented with the promise is more elegantly done with fetch? Let me try again to redo with the fetch (and now I see that you are putting ALL the data in one form field, which makes sense!) Are there size limits on that?

```javascript
/*
 * Expfactory.js
 * Helper module for static experiments to submit data to a server, intended
 *        for use with the Experiment Factory Reproducible Containers.
 *        https://expfactory.github.io. For the expfactory server, the header
 *        before the request is expected to have an ajaxSetup that adds the 
 *        csrf_token (and this is handled by the server template)
 *
 * Requires:
 *     jquery
 *   
 */


function Expfactory () {

    this.data = [],

    // getData
    this.getData = function() {
        return this.data;
    }

    // Add data
    this.addData = function(data) {
        this.data.push(data);
    }

    // Local save function, will just print to console
    // unless the user writes own function (recommended)
    this.localSave = function(data) {
       var data = data || self.data;
       console.log(data);
       console.log(err);
    };

    // Submit Data /save
    this.saveData = function() {
        console.log("Saving data...");    

        // Serialize the data
        var promise = new Promise(function(resolve, reject) {
            var data = JSON.stringify(this.data);
            resolve(data);
        })

        promise.then(function(data) {
            $.ajax({ type: "POST",
                     url: '/save',
                     data: { "data": data },
                     success: function(){ this.nextEndpoint(); },
                     dataType: "application/json",
                     
                     // Endpoint not running, local save
                     error: function(err) {

                         if (err.status == 200){ document.location = "/next"; } 

                         // Do local save of the data
                         else { this.localSave(data); }

                     });
                 })
             }
    }

    // Get data for inspection
    this.getData() = function() {
        return this.get('data');
    }

    // Akin to hitting /next
    this.nextEndpoint = function() {
        console.log("Next experiment!");
        document.location = "/next";
    }

};

```

**Comment by vsoch** [comment]
I want the user to be able to create an experiment, and then optionally to customize the local save for the experiment, so it could be served statically from a Github repo *and* plug into expfactory. So if I remember the saving way for jspsych, for example, they could do like:

```javascript
experiment = Expfactory();
experiment.localSave = function(data) {
    jsPsych.data.localSave('spatial-span_results.csv', 'csv');
}
```
and it would be called still with the expfactory.js integration!

**Comment by FelixHenninger** [comment]
> Is there any reason it has to be form data?

That's a fantastic question! In principle, I don't think there is (by default, the `lab.js` transmission sends just the JSON-encoded data), but the backend [assumes that the data is encoded in fields](https://github.com/expfactory/expfactory/blob/master/expfactory/views/main.py#L118), and will raise a `400` error if it isn't. I don't know how jsPsych transmits data (if it offers that functionality in-library at all), if it uses jQuery internally it probably sticks to their default.

There's probably pros and cons to this approach. I'm guessing jQuery does it that way because of graceful degradation -- it's has the same effect as filling an `HTML` `<form>` tag with the data and submitting it, so you could use the same backend to collect pure form data without any script involved (though I guess that's not a likely scenario for JS-based psychophysics experiments). The con is probably that it's slightly more verbose in as pure JS, as you've noticed.

**Comment by FelixHenninger** [comment]
That's super-cool! A few quick remarks (that would probably be better discussed over in your repository, but since we're here):

* You wouldn't need to wrap the JSON parsing in a promise in [line 42](https://github.com/expfactory/expfactory.js/blob/master/expfactory.js#L42) -- it's synchronous, so it'll wait for you; you could put it just [in front of the AJAX call](https://github.com/expfactory/expfactory.js/blob/master/expfactory.js#L49)
* I don't think you necessarily need ajaxSetup -- the `fetch` example above transmits the CSRF token too!
* I think that in your examples, you might need to write `var experiment = new Expfactory();` to generate a new instance; I'm not sure what happens without the `new`, because the function doesn't return anything.
* w/r/t the API, I'm wondered for a bit whether there is a more composable way of setting up the fallback function (compared to setting it as a class property via `expfactory.localSave`). But the only alternative I can think of would be to return a promise from `expfactory.saveData`, and letting the library catch problems, which doesn't seem as nice.
* This is a purely cosmetic remark, but I think you might enjoy the JS [class syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes). In my eyes at least it's Python with curly braces, so I think you might find it more intuitive (as I did). I've gotten so used to it that I had to look up how function-based OOP worked :-) . Again, this is [not IE11 compatible](https://caniuse.com/#search=classes), but I can help you set up a transpiler that will output backwards-compatible code if you like.

**Comment by FelixHenninger** [comment]
Let me know if you could use a pull request on any of these, I'm happy to help but wouldn't want to get in your way!

**Comment by FelixHenninger** [comment]
Another thing I've been thinking about is how to handle the `CSRF` token -- right now, I think you append it to the experiment's `HTML` code, right? As an alternative, you could set it up in a cookie when you serve the experiment, and extract it from there within `saveData` -- or even better, you could set it when you serve the experiment, and it should automatically be transmitted back with any `fetch` request. That way, you wouldn't need to modify the study at all.

**Comment by vsoch** [comment]
@FelixHenninger I didn't take that approach because (historically) when testing with actual users, most just block cookies entirely (and then nothing works) so this approach to add the header with ajax to the html --> https://github.com/expfactory/expfactory/blob/master/expfactory/templates/experiments/experiment.html is the one that I chose to do!

Could you elaborate more on "set it up when you serve the experiment?" The idea is that ANY static experiment that submits to the endpoint should just work (without needing to do much special).

I'll take a look at the other points above in a few minutes and update the expfactory.js for you to look at!

**Comment by FelixHenninger** [comment]
> when testing with actual users, most just block cookies entirely (and then nothing works)

That's interesting, and makes total sense! I thought I was the only one ;-)

That being said, I just tested running a study without cookies, and it can't get beyond the setup screen -- so there does seem to be a dependency on them somewhere, maybe in Flask's session management?

> Could you elaborate more on "set it up when you serve the experiment?" The idea is that ANY static experiment that submits to the endpoint should just work (without needing to do much special).

Ah, no, my idea would have been cookie based -- you could have set up the `CSRF` cookie in the headers when you send the experiment `HTML` to the browser, and then the cookie would be sent back with the data `POST` request automatically; so the experiment would 'just work' without modifying the `HTML` file at all. (I think that's what happens

**Comment by vsoch** [comment]
> That being said, I just tested running a study without cookies, and it can't get beyond the setup screen -- so there does seem to be a dependency on them somewhere, maybe in Flask's session management?

Most likely - and I'm not sure how to get around that, but let's call it a bug for now and (hope) that sessions can be better / differently handled at some point.

The original experiment factory (version 1.0, still powering my old lab at http://expfactory.org) used cookies, and running on MTurk for (at least hundreds, maybe many more now) of sessions gave us some good insight that a lot of people use some variation of Windows + Internet Explorer, and block cookies. The biggest bug that came in / issues that came up were always around that. So for the second time around, I decided to try and avoid cookies as much as possible, although it seems that I completely forgot about sessions, so we are in the same boat likely of needing to instruct the user to not disable them. That said, maybe there is some difference - it looks like Flask uses secure (signed) cookies (so they can't be changed) https://stackoverflow.com/questions/38427923/how-does-flask-sessions-work and probably this is different from ones we would set with a javascript library (or other) because it's done by the server? Anyway, nothing is seriously broken or not functioning, so we aren't in such a terrible boat. I do think that there could be a lot of improvement, and I am excited to work on this in the time that I can allocate for that!

**Comment by vsoch** [comment]
see here --> https://github.com/expfactory/expfactory-docker/blob/master/expdj/apps/experiments/templates/experiments/serve_battery_cookie.html that's the template that would grab the csrf token cookie! Unfortunately it would lead to tons of errors because it wouldn't be set.

**Comment by FelixHenninger** [comment]
> [R]unning on MTurk for (at least hundreds, maybe many more now) of sessions gave us some good insight that a lot of people use some variation of Windows + Internet Explorer, and block cookies.
Oh, I can totally see that! I'm glad I just have to deal with frontend stuff 😓. Sorry, I think I'm not expressing my self particularly well here -- I didn't mean to criticize your approach, which makes total sense (and more so in light of your experiences and requirements).

**Comment by vsoch** [comment]
Oh I didn't take it as criticism at all! I'm just sharing experience, and enjoying the conversation quite much :)

**Comment by FelixHenninger** [comment]
Oh, thanks, that's good to hear! Very much the same here :-) . Is there some way I can make myself useful beyond idle commentary?

**Comment by vsoch** [comment]
okay some updates for you!

 1. I removed the promise, and put the JSON.stringigy in front of the ajax call.
 2. For fetch - note that I am not using it in expfactory.js. Does it hurt to have an ajaxSetup alongside a fetch? I would want the older experiments (that use ajax and ajaxSetup) to still work, even if a newer experiment chooses to use fetch (such as LabJS). I'm hesitant to switch the expfactory.js over to using fetch because it requires the header token to be written into the experiment. This would break given any use outside of the experiment factory container, which goes outside of the philosophy a bit. But if it's the case that 1. leaving the ajaxSetup in the template (so the older experiments don't break) and 2. an expfactory.js with fetch *wouldn't* break given the csrf_token templating, then I'm definitely in support of this! Would you be able to tweak the expfactory.js that I've created, and create a folder experiment that I can install for us to test?

The important thing is that with ajaxSetup I can put the addition of the csrf_token outside of the javascript, whereas with fetch I can't do anything like that. Are you sure that this ajax setup not work with LabJS, period?


> I think that in your examples, you might need to write var experiment = new Expfactory(); to generate a new instance; I'm not sure what happens without the new, because the function doesn't return anything.

Done! This is me not knowing JS well :)

> w/r/t the API, I'm wondered for a bit whether there is a more composable way of setting up the fallback function (compared to setting it as a class property via expfactory.localSave). But the only alternative I can think of would be to return a promise from expfactory.saveData, and letting the library catch problems, which doesn't seem as nice.

What do you mean composable? I probably didn't do that right - it's how I would do something with Python, for example, just for a subclass to override a superclass function.

> This is a purely cosmetic remark, but I think you might enjoy the JS class syntax. In my eyes at least it's Python with curly braces, so I think you might find it more intuitive (as I did). I've gotten so used to it that I had to look up how function-based OOP worked :-) . Again, this is not IE11 compatible, but I can help you set up a transpiler that will output backwards-compatible code if you like.

I saw that! I started implementing with classes, but then when I scrolled down [the page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#browser-compatibility) I noticed that Internet Explorer doesn't have support for classes:

![image](https://user-images.githubusercontent.com/814322/42728490-43f89100-8770-11e8-901a-10f66c920514.png)

And unfortunately I would guess that a large subset of users (with slightly older computers) are still using that with Windows, so I did a massive backspace (delete) and started over the "old school" way. Is it an old school way now?

What I'd like to suggest is the above - we have two goals at the moment - 

 - get a version of labjs working in this setup
 - provide an expfactory.js for use to make this easier.

The first, not considering the second, would primarily be on your end, to make tweaks to the labjs code, export a test experiment, and then I would rebuild the container (for both of us to test!)

But I think you were hinting that instead of making those changes, this expfactory.js could be used (and you would add the script and use it to your export). I like this idea too, and if that is the best approach then we should do that first. In other words:

 1. Take a look [here](https://github.com/expfactory/expfactory.js) and submit a quick PR to have the post be what you are suggesting, an approach that would be able to exist with ajaxSetup. Or copy paste here and tell me exactly what to do, and I'll try (and then you can check over).
 2. Once the above is done, then if you show me an example of what the rendered thing would look like with an experiment, I can rebuild the container for us to test. Note that the expfactory.js would need to be bundled with the experiment, so that if someone found just the static folder (outside the container, on the repo's Github pages) it wouldn't break because it doesn't find the file.
 3. Once we've both tested the example container, then you might do your magic to have the actual export spit out what we generated.
 4. Then I'll use that export to (again) build the container, and verify it works.

And then *magic* because I can write up the tutorial in full to add to the LabJS docs (officially) and we can share, with confidence that it works from export --> build --> deploy --> use!

**Comment by FelixHenninger** [comment]
Hej, that looks fantastic! A few armchair code review notes, as before -- please don't hesitate to say 'why don't you implement that yourself, Felix, you lazy bum' at any time ;-).

> For fetch - note that I am not using it in expfactory.js. Does it hurt to have an ajaxSetup alongside a fetch?

I think these are interchangeable -- `ajaxSetup` looks jQuery-specific to me, so it'll work alongside `$.ajax`. You could use the combination of those two, or `fetch` as an alternative.

> I'm hesitant to switch the expfactory.js over to using fetch because it requires the header token to be written into the experiment.

Ok, I'm not 100% sure I see what you mean here -- you would append code to a study in any case, right? So the concern is that an experiment would use `$.ajax` internally, and would need to be modified to take into account the `csrf_token` variable, is that correct? If the already existing experiments depend on having the ajax setup for them, then it makes total sense to keep `jQuery` in place. Sorry, I didn't see that dependency.

One thing you might do is check whether `jQuery` is active in the code you append to the study, and only run `ajaxSetup` if it is (i.e. if the `$` variable is set), and rely on fetch in `expfactory.js`. That way, new studies could use your new library, while old ones with `jQuery` would still receive ajax settings?

> This would break given any use outside of the experiment factory container, which goes outside of the philosophy a bit.

So you'd like to run the same experiment, unmodified, both within the container and stand-alone? What role would `expfactory.js` play in that context?

> The important thing is that with ajaxSetup I can put the addition of the csrf_token outside of the javascript, whereas with fetch I can't do anything like that. 

How about a function `expfactory.addHeaders({})` (or, for that matter, a `headers` property on the `Expfactory` class) where you can pass in the headers that are sent along with the request? That would basically work just like `ajaxSetup` (it might even call `ajaxSetup` internally if you choose to use `jQuery`)?

> What do you mean composable? I probably didn't do that right - it's how I would do something with Python, for example, just for a subclass to override a superclass function.

Oh no, sorry, that was me thinking aloud -- I think what you've done is perfect; I was just thinking about whether you could avoid the callback function, but I don't think there's a way. The alternative I was considering was to return a promise from `saveData`, and leave it to the experiment code to catch failures. But that places a lot of responsibility on the study code, and that's probably not a good idea -- calls to `saveData` should probably be fire-and-forget.

One (more constructive) thing you might consider is renaming the callback to something like `onTransmissionError` or `onError` maybe -- researchers might not want to save data locally even if the transmission fails. But that is maximum bikeshedding :-)

> I started implementing with classes, but then when I scrolled down the page I noticed that Internet Explorer doesn't have support for classes

Yeah, I know (sorry, I made a cryptic note in passing above, I should have expanded on that). What's often done is that the raw JS you write goes through a transpilation step before being uploaded somewhere.

In the same step, `fetch` (if you chose to use it) would need to be exchanged by lower-level code that is `IE11`-compatible, but that step is also fairly common and well-supported in the JS world.

With `lab.js`, we use classes and many other advanced JS features that are only slowly being adopted by browsers, but still maintain compatibility because the code is automatically optimized and made backwards-compatible in the build step. That way, we get the best of both worlds, and I get to maintain both my sanity and a 10kLOC codebase ;-) . That's probably actually one of the biggest differences compared to jsPsych; I think our codebase will be much easier to maintain in the future, and it's much easier to test automatically, too (we have >90% test coverage for the core library).

> What I'd like to suggest is the above - we have two goals at the moment -
> 
> * get a version of labjs working in this setup
> * provide an expfactory.js for use to make this easier.

Totally agree, thanks for structuring this!

> But I think you were hinting that instead of [updating the `lab.js` export directly], this expfactory.js could be used (and you would add the script and use it to your export).

That was exactly what I was hoping -- the alternative is that I add support for form-based transmission to `lab.js`, or I can actually do that in any case.

> Take a look here and submit a quick PR to have the post be what you are suggesting, an approach that would be able to exist with ajaxSetup.

Sorry, here I think I lost you -- could you briefly clarify the requirement of being able to exist with `ajaxSetup`?

> Note that the expfactory.js would need to be bundled with the experiment, so that if someone found just the static folder (outside the container, on the repo's Github pages) it wouldn't break because it doesn't find the file.
Ok, I'll need a moment to think about this; from a `lab.js` perspective, we'd probably advise users to export the experiment with an expfactory integration of they want to run it in an expfactory container, and with local data download if it's going to be run independently. But I see your reasoning here.

**Comment by vsoch** [comment]
> So you'd like to run the same experiment, unmodified, both within the container and stand-alone? What role would expfactory.js play in that context?

Yes, the experiments shouldn't be only for experiment factory, the idea is that a static thing (outside of expfactory or labjs or other) could be plugged in, given minimal requirements of having the POST to /next, etc. expfactory.js isn't needed, it would just be a helper that is provided if someone wants to use it (and doesn't want to implement the actual calls to do the posts). That's why it is generally so simple...

> How about a function expfactory.addHeaders({}) (or, for that matter, a headers property on the Expfactory class) where you can pass in the headers that are sent along with the request? 

Flask works by rendering the html file as a template, not the javascript. This wouldn't work to add the csrf_token, unless I required the user to explicitly write the csrf_token variable, which would then break the static experiments. I hope you see what I'm getting at...

> That would basically work just like ajaxSetup (it might even call ajaxSetup internally if you choose to use jQuery)?

I think it might, but it would require the expfactory.js to be rendered as a template, which isn't an option. Actually, it would be an option given that the entire library were embedded in the experiment template. But then if the experiment that the user provides references expfactory.js, it wouldn't be there. This would *only* work given that expfactory.js is provided by the static template (without expfactory) AND embedded in the template, and then we would need to worry about ordering. It's overly complex, and the reason I'm not advocating for an approach that requires the experiment itself to do any kind of thinking about setting the header.

> Oh no, sorry, that was me thinking aloud -- I think what you've done is perfect; I was just thinking about whether you could avoid the callback function, but I don't think there's a way. The alternative I was considering was to return a promise from saveData, and leave it to the experiment code to catch failures. But that places a lot of responsibility on the study code, and that's probably not a good idea -- calls to saveData should probably be fire-and-forget.

Exactly! I want to place minimum requirements on the experiments (code).

> One (more constructive) thing you might consider is renaming the callback to something like onTransmissionError or onError maybe -- researchers might not want to save data locally even if the transmission fails. But that is maximum bikeshedding :-)

good idea, done!

> Yeah, I know (sorry, I made a cryptic note in passing above, I should have expanded on that). What's often done is that the raw JS you write goes through a transpilation step before being uploaded somewhere.

Holy cow, yeah I really don't know how this works!

> In the same step, fetch (if you chose to use it) would need to be exchanged by lower-level code that is IE11-compatible, but that step is also fairly common and well-supported in the JS world.

:O 

> With lab.js, we use classes and many other advanced JS features that are only slowly being adopted by browsers, but still maintain compatibility because the code is automatically optimized and made backwards-compatible in the build step. That way, we get the best of both worlds, and I get to maintain both my sanity and a 10kLOC codebase ;-) . That's probably actually one of the biggest differences compared to jsPsych; I think our codebase will be much easier to maintain in the future, and it's much easier to test automatically, too (we have >90% test coverage for the core library).

I know, it's so beautiful :) I must say I don't know enough JS to be able to create this on my own. I know you must be busy, but I would be *so so* appreciative if you could do a PR with this implemented exactly as you envision. You are the better JS programmer, and I want to do what you think is best, and I'll learn, test, and then comment / ask to discuss what doesn't work given the new implementation. How does that sound?

> Sorry, here I think I lost you -- could you briefly clarify the requirement of being able to exist with ajaxSetup?

Probably you can disregard this for now, and I'll just test the new approach and then bring it up again if there is issue. Basically, the ajaxSetup will still be inserted by expfactory, but if your experiment (or generally, an experiment using expfactory.js) isn't posting with ajax, it shouldn't be an issue (fingers crossed?)

> Ok, I'll need a moment to think about this; from a lab.js perspective, we'd probably advise users to export the experiment with an expfactory integration of they want to run it in an expfactory container, and with local data download if it's going to be run independently. But I see your reasoning here.

okay. please think about this! This is one of the reasons I didn't want to add external dependencies (in terms of files) - it makes it complicated because it changes from "just post to these endpoints with some data structure" to "well, you need to bundle with this extra file." This is why I was going for a "this is the simple code you could adjust the post endpoint for to include" and then plugging in to expfactory would just work (given the ajaxSetup).

You do have the custom file to export expfactory, so I would say a good approach would be to *not* add any extra expfactory.js, but have a function (or added variables to existing ones) that can further customize the behavior of the exported experiment. This way, as other platforms come in to use and export labjs, they can use the same functions but customized to the places they are intended to go. What I don't think you want to be in the business of doing is needed to also serve some custom set of files for the export of choice. So, after all this discussion of expfactory.js, I'd say in practice to throw it away (but we can develop it to still provide to other users that might not want to think about saving and submitting data for an experiment, and don't mind adding the extra file).

**Comment by FelixHenninger** [comment]
Oh, and totally unrelatedly: Do you visit any conferences that a psychologist might find an excuse to travel to? I would love to brainstorm/discuss some ideas in person, and I can't wait 'til 2021 when the Psychonomics swings by California next time ;-)

**Comment by FelixHenninger** [comment]
Sorry, missed your comment, reading now!

**Comment by vsoch** [comment]
In the last couple of years I have been mostly going to research computing / HPC conferences, but I'm infamous for not taking any kind of vacations (and my hours are adding up and need to be used) so it might be a good idea (and really fun!) to take a trip to a conference, and take the opportunity to connect with the community about Experiment Factory. It was a project that I feel rather strongly about (well, this is probably obvious) that I never got to continue working on after graduation, at least officially.

What conferences do you think are good / you would visit, and that would possibly be a good venue to present or connect with the community? Would an event hosted by Mozilla or Open Science Framework be better? I'm open to any and all things, and even willing to spend some of my own money on travel. If you have a proposal, I'm definitely down to do a bit of development, writing if needed, and then a fun trip!

**Comment by vsoch** [comment]
Awesome let me check out this list! And please take your time on (the other not as easy?) responses above - I'm calling it for the day in terms of work, and mostly will be enjoying dinner and reading the internet :) Chat tomorrow / later, and I'll ping you after looking at these fantastic links!

**Comment by FelixHenninger** [comment]
> Yes, the experiments shouldn't be only for experiment factory, the idea is that a static thing (outside of expfactory or labjs or other) could be plugged in, given minimal requirements

Ok, having thought about this for a while, I think I understand now what you're going for, and why I'm slow on the uptake. I have a few ideas, but my hunch is that these are better discussed outside of GitHub issues. Can you spare a bit of time for a brief discussion via skype/hangout/... over the next days, or would you be interested in joining the [`lab.js` slack](https://slackin-nmbrcrnchrs.herokuapp.com/)?

**Comment by FelixHenninger** [comment]
> I'm calling it for the day in terms of work, and mostly will be enjoying dinner and reading the internet :)

Sorry, simultaneous posting again! That sounds awesome, have a great evening! I'll keep on thinking about this, and will (hopefully) catch you around sometime this week.

**Comment by vsoch** [comment]
omg the deadline is tomorrow?!? New Orleans is really close so logistically it's not hard - but I would need to submit!!

**Comment by FelixHenninger** [comment]
It's only an abstract :-) (I also still need to write up my submission)

**Comment by vsoch** [comment]
What if I wrote up the little bit on the experiment builder? Am I allowed tomorrow to write it to (e.g., through the 18th?) That would probably be my one shot :)

**Comment by FelixHenninger** [comment]
I am absolutely 100% certain that that would fit right in! (the bar for SCiP isn't exactly high, either -- please do take a look at last year's programme if you haven't already, to check whether it's worth you while). 

> Am I allowed tomorrow to write it

I ~~very much hope so~~ am pretty sure -- I have it in my calendar for tomorrow. I wish I had thought about this earlier, we could have put together a kick-ass symposium!

**Comment by FelixHenninger** [comment]
Ok, so back to the issue, there's a temporary fix in 24fe0c544ccfbbdf67404f897e079e621d84f1c0, and I've made a pull request back to your repo so. that the changes can make their way into the container for further testing.

I'm still very much in favor of a less close coupling/abstraction layer, but the current solution looks fine, and I hope it works, too!

**Comment by vsoch** [comment]
@FelixHenninger you are so speedy! I'll do a proper review of these things today and post an update.

**Comment by FelixHenninger** [comment]
Yay, thanks, looking forward to that! Do take your time; if you have a pointer for local testing, I'd be glad to try things myself. Have a great Sunday!

**Comment by vsoch** [comment]
Here I will record my complete notes for testing, for @FelixHenninger and actually for myself too, for future help if needed!

1. First, let's understand what happens when we merged. The container is built automatically via the .circleci/config.yml file in the repository, which sets up a build -> test-> deploy pipeline [here](https://circleci.com/workflow-run/ab4cf57a-6977-4507-a0ea-7838ba220f43). Note that I have now added the link to the workflow in the main README, and I also realized that we weren't running the testing pipeline given a PR (and this is also now in place, see the example [here](https://circleci.com/workflow-run/f23a248d-cc03-4c8c-8fe0-ded33fd82006)) @FelixHenninger the PR that you submit should have been tested in this way too. We can't actually deploy with a testing container (the circle docs say sharing credentials isn't secure) so for the future we would need to have the testing happen after build on circle. 

From both of the above avenues (a merge into master) the workflow will run to pull a container to docker hub that we can test. 

```bash
docker pull vanessa/expfactory-stroop:latest                              # <-- merge into master
```
For the remaining instructions, let's pretend we just opened the second pr, number [https://github.com/expfactory/builder-labjs/pull/4](https://github.com/expfactory/builder-labjs/pull/4), we tested and approved it, merged it to master, and are now pulling the container to test locally. Note that for some future version of the builder, more testing would happen on CircleCI via selenium, etc.

## Pull to Local Machine
Since I have the image locally, I need to pull it "fresh" to get the updated version. The pull itself will tell me this by showing new container image downloaded (see last line):

```bash
$ docker pull vanessa/expfactory-stroop
Using default tag: latest
latest: Pulling from vanessa/expfactory-stroop
b234f539f7a1: Already exists 
55172d420b43: Already exists 
5ba5bbeb6b91: Already exists 
43ae2841ad7a: Already exists 
f6c9c6de4190: Already exists 
0706f3f9ff3c: Already exists 
36e2ac494b7d: Pull complete 
2211c4555dfe: Pull complete 
b09e837f2a80: Pull complete 
6ad6188f50fd: Pull complete 
9eee6ce140d5: Pull complete 
f1d96a13fe5c: Pull complete 
391e8dcc0dba: Pull complete 
99a1e2c1ebef: Pull complete 
359acc506ce2: Pull complete 
954600b56110: Pull complete 
7778f65d3e0b: Pull complete 
2926405f1a7e: Pull complete 
2dcaa7008839: Pull complete 
bdd0f389a3ea: Pull complete 
Digest: sha256:e9da0f4bc54df9bbe3d11a2c47faffe0aefc14b273d4b73c71b85c151fcdaf83
Status: Downloaded newer image for vanessa/expfactory-stroop:latest
```
Now we have the entire experiment in it's container, ready to go! We just need to run it. I'm going to give it a name so I can also stop it easily:

```bash
docker run --name stroop -p 80:80 vanessa/expfactory-stroop start
```

With the command above, it's going to be interactive so that pressing Control +C would kill it. If you didn't want this (a "detached") container, you would add `-d`

```bash
docker run --name stroop -p 80:80 -d vanessa/expfactory-stroop start
```
and then you would stop and remove like this, when you are done:

```bash
docker stop stroop
docker rm stroop
```
Once the container is running, open your browser to [http://127.0.0.1](http://127.0.0.1) to test. Here is a shoddy screenshot of my terminal (showing the interactive gunicorn server) and the localhost:

![image](https://user-images.githubusercontent.com/814322/42736916-ddd05152-8822-11e8-9c26-3724461bad94.png)

I'm still testing (it didn't work off the bat and I'm trying adding jquery again) so please hold off testing until I've been able to see if I can fix it up! I'm posting this now out of fear of closing the tab and losing my notes entirely :P

**Comment by vsoch** [comment]
I'll show also how to test by shelling in - I am rebuilding with jquery first. Then I think this would also be useful to have in docs (how to test locally) so I can add there too. And oups I need to write that abstract too, maybe should do that first :_)  Stay tuned!

**Comment by FelixHenninger** [comment]
Woah, thank you, that's a fantastic resource! I've also been hooking into the running container to figure things out -- looking forward to building things properly!

**Comment by vsoch** [comment]
Hey @FelixHenninger I have prepared a document that will walk through the exact steps I took to test this, and I have some things for us to look at together that I think might shed some light on what is going on. First, [here is the repo](https://github.com/expfactory/expfactory-testing) (but no need to go here yet), I will share what I've learned so far!

## 1. Is Jquery influencing it?
The first thing I wanted to know is if the jquery/not jquery being present made a difference. I found that when I had vs. didn't have it, it didn't make a difference. I got 400 error on /save each time. What I didn't test is removing the ajaxSetup because this would break the remainder of experiments. Just to be sure that it has absolutely nothing to do with ajaxSetup, I also rebuild and ran the entire thing removing it (and preserving your version). This didn't make a difference - the 400 was still there. I added it back in to continue testing.

## 2. Is it the format of the data or headers?
I wasn't actually sure about the data structure where the data is being posted, or how it's working. I went back to what you showed me above to do the post. I remembered that you got ajax to work too, so I wanted to verify this for a sanity check

```bash
$.ajax({
  type: "POST",
  url: '/save',
  data: { "data": study.options.datastore.exportJson() },
  dataType: "application/json",
  success: function() { console.log('success') },
  error: function(err) { console.log('error', err) }
});
```
Yes, it worked! So maybe if we compare this working POST to the one that is being done with fetch, we can figure out the issue? Here is the broken one with fetch:

![image](https://user-images.githubusercontent.com/814322/42739205-6cdfa162-8847-11e8-9f11-42059b33da78.png)

and here is the working one, with ajax:

![image](https://user-images.githubusercontent.com/814322/42739195-48631ae4-8847-11e8-9462-87cac679eb54.png)

I think that the answer lies here - see how different they are?  Just for reference, here is the CORS setup for the Flask server:

```bash
# Cors
cors = CORS(app, origins="http://127.0.0.1",
            allow_headers=["Content-Type",
                           "Authorization",
                           "X-Requested-With",
                           "Access-Control-Allow-Credentials"],
            supports_credentials=True)

app.config['CORS_HEADERS'] = 'Content-Type'
```

The fact that the 400 bad request is missing something tells us that there is still a difference between what this test stroop-task is posting internally with fetch, and what the ajax does. This suggests that something is missing so the server will accept it period. Speaking of, I tested this too!

## 3. Is the server rejecting it?
Sometimes 400 can be returned by the server (as in, the /save view) so I added some printing / pickle export of variables within the view. If it was the case that the view is hit (and then rejected) I would see the saved files, but I didn't. Which tells us with confidence that the issue is indeed the fetch vs. ajax POST, and something to do with the header.

Is this a good hint to help us debug and go down the right path? I'm going to zoom out of writing and look at those two pictures now - I think the answer is probably hiding there. I hope you can help!

**Comment by vsoch** [comment]
Differences that I see:

 - working has content type `application/json`,  and broken is `text/html`
 - working has a header called `Set-Cookie` which I'm guessing is done by the ajaxSetup?  --> https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
 - both have the X-CSRF-token header, and the values are equivalent, so that's not the issue (I was worried that somehow the definition of the cookie would be different in the two locations).

Could it just be the html/text vs application/json, since the server is accepting application/json? (see next section in pictures) I can try to tweak the ajax post above to see if it then fails...

**Comment by vsoch** [comment]
Actually hang on, I think we need to look at the request (not response) Take a look at the failed payload, here:

![image](https://user-images.githubusercontent.com/814322/42739267-29122c50-8849-11e8-80e5-2bf407e4b5ab.png)

which is finely formatted json, and take a look at the one that was accepted:

![image](https://user-images.githubusercontent.com/814322/42739276-4e287184-8849-11e8-8a69-7d3eb93ccd32.png)

so maybe your intuition about sending form data was spot on?

**Comment by FelixHenninger** [comment]
Waitwait, oh no! That's not what the library's supposed to be doing! Thanks for taking this apart systematically, I think this is all my mistake 😧😓. Argh, please give me a moment, I'll look at this again.

**Comment by vsoch** [comment]
And finally, full request headers (not working first):

![image](https://user-images.githubusercontent.com/814322/42739290-92e9acd4-8849-11e8-9e66-0160395df4a6.png)


and working next:

![image](https://user-images.githubusercontent.com/814322/42739299-a7413c2e-8849-11e8-9a0d-709238eb536e.png)

sorry for all the posting - I think the answer is in the above and with your expertise on these things and tweaking the post a bit, we can get it working! I probably won't be able to write up and PR for the docs today because this has taken most of the day, but hopefully soon after we get this red --> green! Let me know what we can try next...!

**Comment by vsoch** [comment]
okay!!! :D

**Comment by vsoch** [comment]
This is actually really good learning, I didn't know I could look this deeply into a POST from the browser to server, I've been debugging these things rather blindly! Browsers are so cool!!

**Comment by vsoch** [comment]
Nope no worries I can re-build! (this is easier than you having to PR) Give me a few minutes...

**Comment by vsoch** [comment]
Go builder, go! --> https://circleci.com/gh/expfactory/builder-labjs/40 :building_construction:

**Comment by FelixHenninger** [comment]
So, for a slightly longer answer -- I think you've discovered exactly the critical test to apply. My mistake was in bundling the stable library version that didn't yet include the form encoding. And on the way, I also discovered that the data wasn't being JSON encoded within the form, so the backend would have always gotten `[object Object]` as the dataset :-/ .

There's (at least) one final thing that might go wrong, namely that even the form-encoded data is sent with `Content-Type` set to `application/json`, that might still create problems.

**Comment by FelixHenninger** [comment]
Yay, looking out for the build!

**Comment by vsoch** [comment]
okay looks like the workflow is done:

https://circleci.com/workflow-run/e60e96fe-b28d-4ef2-83d6-354dff2bc00d

You should be able to pull the container again, and run it. Here is what I'm doing:

```bash
docker pull vanessa/expfactory-stroop

.... # lots of layers downloaded here

# If you had pull /run it before
docker stop stroop
docker rm stroop
```
Then start it again

```bash
$ docker run -d -p 80:80 vanessa/expfactory-stroop start
```

**Comment by FelixHenninger** [comment]
Hmm, I'm still getting a `400` code, but also `lab.build.commit` return `fa535b`, which is too old :-/ . Maybe I pulled the container too early? I'm gonna try again, and then poke around inside the container.

**Comment by FelixHenninger** [comment]
Ah, I'm getting a new version now! Will retry ASAP.

**Comment by vsoch** [comment]
okay here is what I'm seeing now, it looks very different still - (it's definitely a different container, but the logic for the post is different):

![image](https://user-images.githubusercontent.com/814322/42739553-2030de10-884e-11e8-970b-6eb17e8bfc99.png)

**Comment by vsoch** [comment]
and request headers (the text/html looks fixed!)
![image](https://user-images.githubusercontent.com/814322/42739563-45fb1516-884e-11e8-95fd-1707e403448d.png)

**Comment by FelixHenninger** [comment]
Yes, that what I'm getting too! Looking at whether this works with the backend.

**Comment by FelixHenninger** [comment]
(`lab.build.commit` should now give you `18e747`)

**Comment by FelixHenninger** [comment]
Ok, no data is being saved :-/. I'll take out the `content-type` headers and try again.

**Comment by vsoch** [comment]
okay! If you want to post an updated zip here, I can rebuild and test again! That seems to be speedy since we are optimized for those things :) I'm going to actually head to start making dinner, so please no rush to look at this! It's Sunday, I'm sure everyone else in the world is like, making pancakes and watching TV :)

**Comment by vsoch** [comment]
Also just to show you, I can see when the data is being saved by looking at the server log, here is from the working POST:
![image](https://user-images.githubusercontent.com/814322/42739614-0eac1be0-884f-11e8-9e80-b8e543594edf.png)

**Comment by FelixHenninger** [comment]
Thanks! I've removed the extra headers and now Chrome says the data are form-encoded, but the backend complains that the CSRF header is missing. This is on me for sure, I shouldn't be doing these things late at night :-) . I'll give it another shot, but I might head to bed if that doesn't work.

**Comment by vsoch** [comment]
okay - let's just call it a day and pause for now then, and how about you can take a look later this week with fresh eyes? I don't even want to ask what time it is there, because it's getting later here (just about 8pm and I haven't started dinner yet, but maybe that's just late for US time).

This was so fun today / this weekend, don't worry we will get to the bottom of it! CORS is sort of weird to get working, but it's so incredibly important for the security of applications, etc. I think I've spent many a days just getting simple things working, and I can't claim to completely understand all points. 

In case this is helpful - I've struggled through this before (with various methods to do the post) and you might get better hints in some of these (working) examples:

https://github.com/singularityhub/interface/blob/master/tunel/templates/recipe/build-modal.html#L56
https://github.com/singularityhub/interface/blob/master/tunel/templates/logs.html#L39
https://github.com/singularityhub/interface/blob/master/tunel/templates/plugins/globus/endpoint.html#L106
https://github.com/singularityhub/interface/blob/master/tunel/templates/main/index.html#L109
https://github.com/singularityhub/interface/blob/master/tunel/templates/main/settings.html#L116

sorry if those are all different variations of the same thing! I think it could help because it uses fetch and has the same server setup as expfactory.

could it just be the credentials same-origin thing? I seem to remember it was super simple and I fell to my knees and threw my hands at the sky when it worked... lol.

**Comment by FelixHenninger** [comment]
Wow, thanks for all of those fantastic examples! I've gotten things to work, but I'll need to take another look at things tomorrow -- I wouldn't want to break more things than I fix. Onto the SCiP submission!

[study_export-library@head.zip](https://github.com/FelixHenninger/lab.js/files/2196277/study_export-library.head.zip)

**Comment by vsoch** [comment]
Good luck! I'll test if I have time later tonight and post the same logs here.

**Comment by vsoch** [comment]
Wooo I got green success 200 with fetch!! Here we go!!

```javascript
 fetch('/save', {
                headers: { "X-CSRFToken": csrf_token },
                credentials: 'same-origin',
                data: { "data": study.options.datastore.exportJson() },
                dataType: "application/json",
 	        method: 'POST' })
            .then((response) => response.json())
            .then(function(data) {
               console.log(data);
             });
```
boomshakalaka!

**Comment by vsoch** [comment]
```bash
var text = new SpeechSynthesisUtterance("BoomShakaLaka!"); 
speechSynthesis.speak(text);
```

**Comment by FelixHenninger** [comment]
Hej Vanessa,

thanks for your updates! This is a versioning problem again, I think the zip I exported must have been out of date :-/ . I'm sorry I'm this useless, things will hopefully get better after I catch some sleep!

The fetch code you've posted above is prior to the last fix -- as a last step, the `JSON` headers need to be removed, and the `CSRF` token set. Here's what the code looks like in my copy:

```js
// Encode data
let body, defaultHeaders = {}
if (encoding === 'form') {
  // Encode data as form fields
  body = new FormData()
  body.append('metadata', JSON.stringify({ slice, ...metadata }))
  body.append('url', window.location.href)
  body.append('data', JSON.stringify(data))
} else {
  // JSON encoding is the default
  body = JSON.stringify({
    metadata: {
      slice,
      ...metadata,
    },
    url: window.location.href,
    data: data,
  })
  defaultHeaders = {
    'Accept': 'application/json', // eslint-disable-line quote-props
    'Content-Type': 'application/json',
  }
}

return fetch(url, {
  method: 'post',
  headers: {
    ...defaultHeaders,
    ...headers,
  },
  body,
  credentials: 'include',
})
```

This is supposed to switch between form-encoding without the headers, and `JSON` encoding with headers. So basically, running `study.options.datastore.transmit('/save', {}, { headers: { "X-CSRFToken": csrf_token }, encoding: 'form' })` should trigger the same command that you put together manually above. This works for me, but I'm probably on some unicorn build here 🦄 .

Thanks for all of your patience, and sorry for my continuous screw-ups!

-Felix

**Comment by FelixHenninger** [comment]
But having written all of that, it's weird that fetch worked for you without the form encoding 🤔. Ok, that's a mystery for tomorrow!

**Comment by vsoch** [comment]
okay, good plan! I think we have figured it out and just need some sleep, so I'm very content with our work today :)

**Comment by vsoch** [comment]
omg it works for me too! I didn't even open the JavaScript console :)

![image](https://user-images.githubusercontent.com/814322/42769332-bf77eee0-88d6-11e8-97e8-6f3e985ddf4f.png)

Verified also that the data is saved! We did it!!! :D

**Comment by vsoch** [comment]
I'll get underway on the docs to go from labJS --> expfactory builder starting later today! And Felix if you like having a container built to deploy some set of library experiments, I will be very happy to set these up for you (maybe an organization?) So excited about this!! 

More from me later!

**Comment by FelixHenninger** [comment]
Yay, that sounds fantastic! 🎈💃 🎉 🕺  I'm super-glad we got this to work; thanks for all of your help, and for humoring my late-night uselessness!

I'd be happy to help with the docs, will close this for now and follow your issue instead, if that's alright.

\*updates manuscript to boast about new feature\*
